### PR TITLE
Mod huber loss

### PR DESCRIPTION
**Pull request by khotilov**
Added Modified Huber Loss objective "binary:huber" for binary classification that is similar to `distribution = "huberized"` from gbm.
http://en.wikipedia.org/wiki/Huber_loss#Variant_for_classification
http://cran.r-project.org/web/packages/gbm/gbm.pdf#page=7
https://github.com/harrysouthworth/gbm/blob/master/src/huberized.cpp

I saw comparable performance to "binary:logistic" in datasets that I've tried,

### CODE REVIEW

**Comment by tqchen** [comment]
Thanks for the PR. One comment:

The second order gradient presented at x > 1 and x <-1 was correct. But I am not completely sure this could be a safe choice (returning eps in second order gradient) for all the cases

Let us consider a safer alternative, for position x, let us find h(x) such that
```
g(z)  = f(x) + f ' (x) (z-x) + 0.5 h(x) (z-x)^2  >= f(x)
```
where f(x) is our loss function. A safe choice will be always set ```h(x) = 2```. Some derivation may be needed to get a smaller h(x) such that it is tight.

**Comment by khotilov** [comment]
I've just looked through your slides again, and I kind of see why having a lot of opportunities for very small 2nd order gradient could be dangerous. Maybe it's late, but I'm not quite catching why would you want the `f ' (x) (z-x) + 0.5 h(x) (z-x)^2 >= 0` constraint? Or why there would be such freedom possible in 2nd order Tailor expansion?

I'm sorry for the slowness - am not having enough time in days recently...

**Comment by tqchen** [comment]
The reason was not explained in the slide, but here is why:

In general, we do not necessarily need a approximation of the original loss function. It will suffice to optimize an upper bound of the original loss function that has the same value at the current point(delta=0). 

The idea is kind like EM or majorization maximization algorithm. Optimizing the upper bound every time will leads to optimization of original loss function. For well behaved function, usually the 2nd order Taylor was a nice tight approximate upper bound. But in cases like huber, you can find that the Taylor(which was a line) will go below the original loss when we do not constrain the movement, this is why I think we need a more conservative upper bound(or constrain the delta of the move)

Note that Taylor expansion was usually accurate when delta is small

**Comment by tqchen** [comment]
Closing this now since the core code will be moving to new codebase https://github.com/dmlc/xgboost/tree/brick

There is also a plugin system that can plug quick experimental features into https://github.com/dmlc/xgboost/tree/brick/plugin

**Comment by khotilov** [comment]
I might want to revisit this at some point... I've tried some MM-inspired ideas, but the convergence with flat gradient just did not look great. That seems to relate to #691
